@xen-orchestra/fs/src/abstract.js

```diff
@@ -95,26 +148,46 @@ export default class RemoteHandlerAbstract {
     }
     ;({ highWaterMark: this._highWaterMark, timeout: this._timeout = DEFAULT_TIMEOUT } = options)
 
-    const sharedLimit = limitConcurrency(options.maxParallelOperations ?? DEFAULT_MAX_PARALLEL_OPERATIONS)
-    this.closeFile = sharedLimit(this.closeFile)
-    this.copy = sharedLimit(this.copy)
-    this.getInfo = sharedLimit(this.getInfo)
-    this.getSizeOnDisk = sharedLimit(this.getSizeOnDisk)
-    this.list = sharedLimit(this.list)
-    this.mkdir = sharedLimit(this.mkdir)
-    this.openFile = sharedLimit(this.openFile)
-    this.outputFile = sharedLimit(this.outputFile)
-    this.read = sharedLimit(this.read)
-    this.readFile = sharedLimit(this.readFile)
-    this.rename = sharedLimit(this.rename)
-    this.rmdir = sharedLimit(this.rmdir)
-    this.truncate = sharedLimit(this.truncate)
-    this.unlink = sharedLimit(this.unlink)
-    this.write = sharedLimit(this.write)
-    this.writeFile = sharedLimit(this.writeFile)
-
     this._forget = coalesceCalls(this._forget)
     this._sync = coalesceCalls(this._sync)
+
+    this._applySafeGuards(options)
+  }
+
+  _applySafeGuards(options, { toLimit = WITH_LIMIT, toTimeout = WITH_TIMEOUT, toRetry = WITH_RETRY } = {}) {
+    const sharedLimit = limitConcurrency(options.maxParallelOperations ?? DEFAULT_MAX_PARALLEL_OPERATIONS)
+    toLimit.forEach(functionName => {
+      if (this[functionName] !== undefined) {
+        this[functionName] = sharedLimit(this[functionName])
+      }
+    })
+
+    toTimeout.forEach(functionName => {
+      if (this[functionName] !== undefined) {
+        this[functionName] = withTimeout(this[functionName], DEFAULT_TIMEOUT)
+      }
+    })
+
+    toRetry.forEach(functionName => {
+      if (this[functionName] !== undefined) {
+        // adding the retry on the top level method won't
+        // cover when _functionName are called internally
+        this[functionName] = pRetry.wrap(this[functionName], {
+          delays: [100, 200, 500, 1000, 2000],
+          // these errors should not change on retry
+          when: err => !['EEXIST', 'EISDIR', 'ENOTEMPTY', 'ENOENT', 'ENOTDIR', 'SystemInUse'].includes(err?.code),
```

Comment:
the retry condition can be specific to a fs 
please define a retry condition property or method that can be overwritten per sub implementation.